File: typeScript/challenge03.ts
type Inventory = Array<
  { name: string, quantity: number, category: string }
>

function organizeInventory(inventory: Inventory): object {
  const invetoryOrganized: Record<string, Record<string, number>> = {};
  for(const {name, category, quantity} of inventory) {
    if(!invetoryOrganized[category]) {
      invetoryOrganized[category] = {}
    }
    invetoryOrganized[category][name] = (invetoryOrganized[category][name] ?? 0) + quantity;
  }
  return invetoryOrganized
}

const inventory = [
  { name: 'doll', quantity: 5, category: 'toys' },
  { name: 'car', quantity: 3, category: 'toys' },
  { name: 'ball', quantity: 2, category: 'sports' },
  { name: 'car', quantity: 2, category: 'toys' },
  { name: 'racket', quantity: 4, category: 'sports' }
]

organizeInventory(inventory)

// Expected result:
// {
//   toys: {
//     doll: 5,
//     car: 5
//   },
//   sports: {
//     ball: 2,
//     racket: 4
//   }

const inventory2 = [
  { name: 'book', quantity: 10, category: 'education' },
  { name: 'book', quantity: 5, category: 'education' },
  { name: 'paint', quantity: 3, category: 'art' }
]

organizeInventory(inventory2)

// Expected result:
// {
//   education: {
//     book: 15
//   },
//   art: {
//     paint: 3
//   }
// }

File: python/challenge03.py
def organizeInventory(inventory):
    grouped = {}
    for item in inventory:
        category = item["category"]
        name = item["name"]
        if category not in grouped:
            grouped[category] = {}
        if name not in grouped[category]:
            grouped[category][name] = 0
        grouped[category][name] += item["quantity"]
    return grouped


data = [
    { "name": 'doll', "quantity": 5, "category": 'toys' },
    { "name": 'car', "quantity": 3, "category": 'toys' },
    { "name": 'ball', "quantity": 2, "category": 'sports' },
    { "name": 'car', "quantity": 2, "category": 'toys' },
    { "name": 'racket', "quantity": 4, "category": 'sports'}
]

data2 = [
    { "name": 'book', "quantity": 10, "category": 'education' },
    { "name": 'book', "quantity": 5, "category": 'education' },
    { "name": 'paint', "quantity": 3, "category": 'art' }
]

data3 = []



print(organizeInventory(data2))